File: LeetCode/JavaScript/B26-23-Merge k Sorted Lists.js
/**
 * 채점 결과 틀림
 */

// Definition for singly-linked list.
function ListNode(val, next) {
  this.val = val === undefined ? 0 : val;
  this.next = next === undefined ? null : next;
}

/**
 * @param {ListNode[]} lists
 * @return {ListNode}
 */
var mergeKLists = function (lists) {
  nodes = new Array();
  lists.forEach((element, i) => {
    if (!!element) {
      nodes.push([element.val, i, element]);
    }
  });

  const root = new ListNode();
  let cur = root;
  while (!!nodes) {
    nodes.sort((a, b) => a - b);
    [_, idx, node] = nodes.shift();
    cur.next = node;
    cur = cur.next;
    if (!!node && !!node.next) {
      nodes.push([node.next.val, idx, node.enxt]);
    }
  }
  return root.next;
};

File: LeetCode/5회차/B26-23-Merge k Sorted Lists.py
from typing import Optional, List
from heapq import heappop, heappush


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        def merge(hq):
            if not hq:
                return None
            _, idx, node = heappop(hq)
            if node.next:
                heappush(hq, [node.next.val, idx, node.next])
            node.next = merge(hq)
            return node

        hq = []
        for i, node in enumerate(lists):
            if node:
                heappush(hq, [node.val, i, node])
        return merge(hq)
